### DISCUSSION TITLE
Parameter for disabling scheduling

### DISCUSSION THREAD

**Discussion by rishabh643** [apache/airflow#29173] [General]
### Description

Can we have a parameter in airflow that can disable scheduling for all dags that are present?
I know one can disable dags individually or can turn off the scheduler altogether.
But I am specifically looking for a parameter through which we can disable the scheduling of all dags without bringing down the scheduler.

### Use case/motivation

We are creating a Disaster Recovery setup. In which primary site will be up and running.
While the secondary site should be up with minimum replicas and no one should be able to trigger the dags on the secondary.
In order to do so I want to set up airflow on secondary with this parameter enabled. So that no one even by mistake can run enable the dag from the UI and run it.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by eladkal** [comment]
Is the issue that you don't want the tasks to be scheduled or you don't want them to actually run?
If you won't have any workers up then nothing will be executed (except EmptyOperator)

**Reply by rishabh643** [reply]
I am looking for both, but "don't want the task to be scheduled" is a higher priority for us.
Since we are following the active-warm setup for all the services we have, we want the workers to be up and running with the minimum replica.

**Comment by potiuk** [comment]
Additionally - why you do not want to disable scheduler? Seems like a reasonable "parameter" of your disaster recovery scenario where you would not start a scheduler, but everything else will be enabled (and you can switch this particular instance to be "working" by enabling scheduler. 

I think this kind of scenario is great to be handled by the way your deployment is running airlfow and "not running scheduler" seems like it achieves what you want.

I am converting it to a discussion - but if you think there are other reasons why scheduler should be running while not doing scheduling, feel free to elaborate on it.

**Reply by rishabh643** [reply]
The approach that we are following for DR is the active-warm setup. Where on the secondary site I want the scheduler to be up and running with a minimum replica. This is why I am looking for such a parameter.
Regarding why anyone generally goes for an active-warm setup, is to make sure that in the event of a disaster, we don't face any surprises, and the service can start taking requests as soon as the switch to secondary is done.
One of the examples of surprises that I mentioned in the previous statement could be we are trying to bring up the scheduler after the switch, but it's not coming up due to XYZ reasons, leading to increased Recovery Time Objective.

**Reply by potiuk** [reply]
Well. Scheduler has "active-active" mode already - which uses Database Locking. You just start second (or 3rd, 4th) scheduler and you are done. 
I am not sure why would you need "active->warm" if you can have "active <-> active" where you can not only have it running continuously and perform rolling restarts, but you also can use the CPU/memory resources while the other schedulers are running. With "active-> warm" setup you have 50% resources not used all the time.

Is there a specific reason you cannot utilise it, or were you simply unaware that Scheduler has this capability ?

**Reply by rishabh643** [reply]
Active-warm is a strategy that the customer has asked for. In the complete architecture, we are using various other components for which we already use active warm. The idea behind this is on the secondary cluster we will run with just 1 replica and minimum resources. And hence we should be able to save cost.
In general, let's say the customer currently uses 20 GB of Memory and 20 CPU for running their load on primary. Active-active would mean setting up a second site with the same resource. However, in the case of Active warm, the second site could be up and running with 1 GB and 1 CPU. And only if the disaster happens and the switch to secondary is performed, this instance will scale up to the max limit defined.
Can you explain in detail the active-active mode of the scheduler and Database locking? Also please share any documentation that we have for it. Will check.

**Comment by potiuk** [comment]
> Active-warm is a strategy that the customer has asked for

Surely. But i guess you are paid to advice your customer what's best and give feedback on a) what's possible b) what makes sense.

> In general, let's say the customer currently uses 20 GB of Memory and 20 CPU for running their load on primary. Active-active would mean setting up a second site with the same resource. However, in the case of Active warm, the second site could be up and running with 1 GB and 1 CPU

This is not how "warm" works for memory that I know of. The CPU/Memory reason for warm active makes no sense when you look at it in general.

How are you going to limit running Python program for allocating most of the memory in  warm scenario? What is your idea for that ? Most of memory "used" by Python programs is allocated when things "warm up". 

Can you show me an example of program (python or otherwise) that would allocate only minimum memory and when running takes more memory and how you are currently using it ?  I would be courious to see such an example. The assumption that warm system will use very little memory and will be "ready" to work makes very littles sense in general. In pretty much all cases I know "warming up" a system means to let it allocate all memory it needs to run so that it can immediately start running.

Regarding the CPU - there is no problem to limit CPU usage for running insteances. You can also do it manually. But that generally makes little sense. Why do you want to limit the CPU ? Mostly to save energy. But If you have two cases:
* one instance takes 3 CPUS and another instance takes 1 CPU
* both take 2 CPUs

In both cases overall energy consumptions is the same at best. 
And usually it is actually better the more you distribute the load, because parallel systems suffer from Amdahl's law https://en.wikipedia.org/wiki/Amdahl%27s_law where the more cores you use, the bigger synchronisation effects are and your system will always gets less efficient per CPU the more CPUS you have. Same Law applies to parallel machines of course, but if the system is well designed for parallelism of multiple instances (Airflow is) this effect is usually smaller. So Active-warm in order to consume energy is chasing a red herring (and for sure in Airflow's case makes no sense).

> Can you explain in detail the active-active mode of the scheduler and Database locking? Also please share any documentation that we have for it. Will check.

Surely. It's all described in our docs in the chapter aptly named "Running More than One Scheduler" https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html#running-more-than-one-scheduler
Also if you want all the details you can get, here is the fantastic talk from Airflow Summit 2021 (the principles and behaviours explained there are still in operation) that goes into all the details you need https://www.youtube.com/watch?v=DYC4-xElccE

**Reply by potiuk** [reply]
One more comment.

I would really be courious to see some examples.  You mentioned you already use it. I would be really curious to see where you are using active-warm and you actually save memory and CPU (rather than only make use of the fact that the system is designed from grounds up to support active-warm (or also similarly named active-passive) because it simply cannot handle active-active. 

Good examples that I know about when it comes to active-passive are a number of databases (Oracle, MySQL, Postgres). But the reason they are active-passive is that their architecture does not allow active-active and they have to have single node to perform all the operations and the other node to consume the results of those operations and "catch up" - ready with all the data in memory and disk to start right after the first fails. This is not to serve resources, but to handle the case that it is super-complex and slow for those systems to run active-active settings. But this is NOT the case for Airlfow we handle Active-Active and this is how Airlfow should be run in multi-scheduler setup.

And also for more context. As opposed to  Airflow scheduler, the database of Airflow is best to run Active-Passive mode if you want to keep high availability. Airflow instances are not capable of talking to two database replicas at the same time, so Active-Passive DB is the only setting.

**Reply by rishabh643** [reply]
The section that you have referred to is for HA. I am looking for Disaster Recovery. 
In the case of HA, there could be multiple instances of the scheduler running in the same region, but different AZs. Here obviously active-active makes sense.
I am referring to a scenario where an instance of the scheduler is also deployed on a different region altogether (for example primary on East US and secondary on West US).
Also, we are using Managed Postgres, Azure Flexible Server for metadata. It provides a read-only replica on the secondary region. This means airflow will not be able to write anything in the DB. Then how will it support active-active?

When I mentioned I intend to save resources. I didn't mean it, particularly for Airflow, this is why I mentioned: "in general". 
This was for our application microservices that are running on secondary.
We basically deploy our services on both primary and secondary regions and make sure that the ones deployed on the secondary do not process anything.
When they do not process anything, HPA makes sure that there is just 1 instance (pod) running for the MS. Hence we save resources, on secondary. When the switch to secondary happens these MS starts processing data, and if the load is high, the HPA will start creating multiple pods.

**Comment by potiuk** [comment]
Airflow does not support this scenario (for the reasons explained above). You won't be able to achieve it.

**Reply by potiuk** [reply]
Just to provide some more thoughts.

I see why you want to do it. But I personally believe the complexity of such an approach is absolutely not justified - when you look at benefits of it vs. costs. And if you want to have such a feature in airflow - even if you get the community to agree to it, which is highly unlikely - you would have to bear the cost of desiging and implementing it yourself, I am afraid.

The reason you mentioned "to prevent the case that new airflow instance might fail to start" is an extremely weak argument. There are multiple reason why such "warmed" instance (if put to "sleep" ) might also fail and crash after it wakes up and complexity of such switch-over is such that it is most likely something goes wrong when you "wake up" rather than when you "start application from the scratch" is much higher. It's a red herring you are chasing again. If you keep same image, same DB repllca, same configuration on both instances - there is no reason restarting Airlfow will be more risky than "waking it up" if we even had this feature.

Airflow has never been meant to have 99.999 availailabily or so (and as essentially batch scheduler it does not have to have such extreme availability). Designing and manging systems that are so available is extremely costly. Such systems are complex and brittle and keeping stuff running and "waking up" is few orders of magnitude more costly than relying on the fact that you can stop and restart the application at any time. So you are looking for high complexity, high cost of implementaiton and high cost of managing it in return for almost no gain.

I guess It would be possible to implement some kind of switchover like that if you really, really really want. I see why you want to do it this way with distributed network but you would have to implement some way of detecting a failover scenario and automatically coordinate switch-over of the active DB instance to the second instance and enable second airflow to schedule. There are probably many other small things -for example you would have to also stop all the maintenance and parsing jobs from running, possibly also somehow reload the ORM connection to the database, find a way to switch and distribute both DAGs and logs and many other small things if you would like to do it preperly. Definitely calls for Airlfow Improvement Proposal if you would like to implement it.

Seems you are knowledgeable about it, and since this is rather niche case and very costly to implement and manaage, but possibly small amount of users would benefit from it, the best approach is to come up with Airflow Improvement Proposal https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals . Most likely (due to niche case and complexity) you would have to lead all the design proposal, discussion and voting and later implementation yourself. 

The way it works, is you start a discussion on our devlist outlining your proposal, see if you have enough consensus on the direction, then you design (in large detail) Airflow Improvement Proposal where you explain what and howyou do (including which parts of the code are affected, configuration needed, discuss consequences and details), then you lead discussion on the devlist and eventually call for voting.. Finally you implemet it as a - usually series of PRs to review and merge. 

However due to arguments I explain before, I guess you will not go pass through devlist discussion phase. We opt for simple rather than complex solutions, and since airflow is essentially batch scheduler, 99.999 availability that you seek has never been our goal. But maybe you will have appealing arguments, who nows. You can always try.

But I would strongly advise you to avoid the high cost and explain your customer that what they are looking for is a red herring.

* Have an active-passive database (cross region)
* Invest in making sure that you have identical deployment on both sides (docker image, configuration etc.)
* Implement a scenario to manually (or automatically if you will to fail-over airflow between the regions - stop airflow in one of them and start in the other

This will be far less costly and problematic, and you will achieve much more stability with it.

**Reply by rishabh643** [reply]
So for our use case of Disaster Recovery, you suggest to setup Airflow on 2 regions. The scheduler on the primary region will be up and running while the scheduler on the secondary region will be down. The Postgres will be in read/write mode in primary and in read-only mode in secondary. Also, we need to replicate the dags directory from primary to secondary. When the disaster happens we bring up the scheduler on the secondary, so that scheduled jobs can run.
Am I missing anything else?

**Reply by potiuk** [reply]
Looks good. I think you also need to swapt the primary database/ replica when you failover. 

But yes. That sounds like how I would do it. Far less complexity and much more robust  solution in general.
